### PR TITLE
sql: add new BACKFILLING mutation state

### PR DESCRIPTION
**Pull request by stevendanna**
This adds a BACKFILLING mutation state. When an index is in this
state, it is ignored for the purposes of INSERT, UPDATE, and
DELETE. Currently, this state is unused in the actual code.

This new state will eventually be used in the new index backfiller. In
the new backfiller, the bulk-backfill using AddSSTable will occur when
the index is in this state, with concurrent updates being captured by
a temporary index. See
docs/RFCS/20211004_incremental_index_backfiller.md for more details.

Indexes in this state are returned by `(TableDescriptor).AllIndexes`, 
`(TableDescriptor).NonDropIndexes`, and `(TableDescriptor).Partial`.

Release note: None

### CODE REVIEW

**Review by rhu713 — APPROVED** [review]

File: pkg/sql/opt_catalog.go [hunk 1]
```diff
@@ -1062,7 +1062,7 @@ func (ot *optTable) WritableIndexCount() int {
 // DeletableIndexCount is part of the cat.Table interface.
 func (ot *optTable) DeletableIndexCount() int {
 	// Primary index is always present, so count is always >= 1.
-	return len(ot.desc.AllIndexes())
+	return len(ot.desc.DeletableNonPrimaryIndexes()) + 1
```
Comment (postamar) [inline pkg/sql/opt_catalog.go:1065]: Why this change? This is the same thing.

**Review by postamar — COMMENTED** [review]
This seems fine on the face of it.

I say on the face of it because there might be code that relies on the assumption that the only possible values are PUBLIC, DELETE_ONLY and DELETE_AND_WRITE_ONLY. However, because the new BACKFILL value is not set anywhere except in that unit test, this assumption is never really challenged.

I'm all for smaller PRs but in this case it would be nice to see this commit in a multi-commit PR followed by other commits which actually exercise this new value. If that's not practical then I'm fine with merging this. Do we have an idea of when we're going to exercise this new value? Generally speaking I'm a bit uneasy with changes "for future use" if it's not clear that this use is in a very-near future.

**Comment by stevendanna** [comment]
@postamar 

> This seems fine on the face of it.
> ..
> I'm all for smaller PRs but in this case it would be nice to see this commit in a multi-commit PR followed by other commits which actually exercise this new value. If that's not practical then I'm fine with merging this. 

Thanks for taking a look! Happy to just fold this into a larger PR. There are a few things that need to come together to make this end-to-end usable.

File: pkg/sql/opt_catalog.go (hunk 1 above)
Comment (postamar) [reply pkg/sql/opt_catalog.go:1065]: I get what you're saying and your change here makes sense in that light. That being said I just pulled `master` I'm not seeing what you describe in the current implementation of `AllIndexes()`. Extending the `indexCache` struct to support backfilling indexes seems like it would lie within the scope of this PR.

Comment (postamar) [reply pkg/sql/opt_catalog.go:1065]: My comment above is incorrect, sorry. You did what was required as far as I'm concerned.

**Comment by stevendanna** [comment]
@rhu713 Mind giving this a second look. I pulled in some of the changes from the integration branch, but otherwise this part of the change hasn't changed in a while and I think it is worth merging this if it still looks good to you.

**Comment by stevendanna** [comment]
bors r=postamar

**Comment by stevendanna** [comment]
bors retry

**Comment by stevendanna** [comment]
The bors error we are getting her doesn't seem to easily reproduce locally for me on a rebased branch and also doesn't happen in pre-merge CI from the looks of it...

**Comment by stevendanna** [comment]
bors retry
